### PR TITLE
ARROW-7261: [Python] Add Python support for Fixed Size List type

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
https://issues.apache.org/jira/browse/ARROW-7261

Conversion from python lists works, but conversion to python (to numpy/pandas) not yet.

### CODE REVIEW

File: cpp/src/arrow/array.cc [hunk 1]
```diff
@@ -447,6 +447,24 @@ std::shared_ptr<DataType> FixedSizeListArray::value_type() const {
...
+  }
+  if ((values->length() % list_size) != 0) {
+    return Status::Invalid(
+        "The length of the values Array needs to be a multiple of the list_size");
+  }
+  int64_t length = values->length() / list_size;
+  auto list_type = std::make_shared<FixedSizeListType>(values->type(), list_size);
+  std::shared_ptr<Buffer> validity_buf;
+
+  *out = std::make_shared<FixedSizeListArray>(list_type, length, values, validity_buf, -1,
```
Comment (emkornfield) [inline cpp/src/arrow/array.cc]: please document, literal parameters.

File: python/pyarrow/array.pxi [hunk 2]
```diff
@@ -1379,6 +1379,54 @@ cdef class LargeListArray(Array):
...
 
+cdef class FixedSizeListArray(Array):
+    """
+    Concrete class for Arrow arrays of a fixed size list data type.
+    """
+
+    @staticmethod
+    def from_arrays(values, int64_t length, MemoryPool pool=None):
+        """
+        Construct ListArray from arrays of int32 offsets and values
```
Comment (emkornfield) [inline python/pyarrow/array.pxi]: copy/paste bug?

File: python/pyarrow/array.pxi [hunk 3]
```diff
@@ -1379,6 +1379,54 @@ cdef class LargeListArray(Array):
...
+        result.validate()
+        return result
+
+    @property
+    def values(self):
+        return self.flatten()
+
+    def flatten(self):
+        """
+        Unnest this ListArray by one level
```
Comment (emkornfield) [inline python/pyarrow/array.pxi]:
```suggestion
        Unnest this FixedSizeListArray by one level
```

File: python/pyarrow/scalar.pxi [hunk 4]
```diff
@@ -731,6 +731,57 @@ cdef class LargeListValue(ArrayValue):
...
+        raise StopIteration
+
+    cdef void _set_array(self, const shared_ptr[CArray]& sp_array):
+        self.sp_array = sp_array
+        self.ap = <CFixedSizeListArray*> sp_array.get()
+        self.value_type = pyarrow_wrap_data_type(self.ap.value_type())
+
+    cdef getitem(self, int64_t i):
+        cdef int64_t j = self.ap.value_offset(self.index) + i
+        return box_scalar(self.value_type, self.ap.values(), j)
```
Comment (emkornfield) [inline python/pyarrow/scalar.pxi]: not familiar with what box_scalar does but it seems strange to be boxing a non-scalar type list?

Comment (jorisvandenbossche) [reply python/pyarrow/scalar.pxi]: > not familiar with what box_scalar does but it seems strange to be boxing a non-scalar type list?

This is getting one element of the list value (not boxing the full list value), which typically is a scalar, unless you have nested lists (so `self` is a single list of the list array, and this method is then getting one element of that list)

**Comment by jorisvandenbossche** [comment]
@emkornfield thanks for the review!

File: cpp/src/arrow/array.cc [hunk 5]
```diff
@@ -447,6 +447,24 @@ std::shared_ptr<DataType> FixedSizeListArray::value_type() const {
...
+  if ((values->length() % list_size) != 0) {
+    return Status::Invalid(
+        "The length of the values Array needs to be a multiple of the list_size");
+  }
+  int64_t length = values->length() / list_size;
+  auto list_type = std::make_shared<FixedSizeListType>(values->type(), list_size);
+  std::shared_ptr<Buffer> validity_buf;
+
+  *out = std::make_shared<FixedSizeListArray>(list_type, length, values, validity_buf,
+                                              /*null_count=*/-1, /*offset=*/0);
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply cpp/src/arrow/array.cc]: Indeed, thanks

File: cpp/src/arrow/array.cc [hunk 6]
```diff
@@ -447,6 +447,24 @@ std::shared_ptr<DataType> FixedSizeListArray::value_type() const {
 
 std::shared_ptr<Array> FixedSizeListArray::values() const { return values_; }
 
+Status FixedSizeListArray::FromArrays(const std::shared_ptr<Array>& values,
+                                      int32_t list_size, std::shared_ptr<Array>* out) {
+  if (list_size < 0) {
+    return Status::Invalid("list_size needs to be a positive integer");
+  }
+  if ((values->length() % list_size) != 0) {
```
Comment (pitrou) [inline cpp/src/arrow/array.cc]: This is obviously problematic when `list_size` is 0. You probably want to add a dedicated code path for that.

File: cpp/src/arrow/array.h [hunk 7]
```diff
@@ -709,6 +709,14 @@ class ARROW_EXPORT FixedSizeListArray : public Array {
     return values_->Slice(value_offset(i), value_length(i));
   }
 
+  /// \brief Construct FixedSizeListArray from child value array and value_length
+  ///
+  /// \param[in] values Array containing list values
+  /// \param[in] list_size The fixed length of each list
+  /// \param[out] out Will have length equal to values.length() / list_size
+  static Status FromArrays(const std::shared_ptr<Array>& values, int32_t list_size,
```
Comment (pitrou) [inline cpp/src/arrow/array.h]: Since we're discussing migrating public APIs to `Result<>`, this one should probably return a `Result<std::shared_ptr<Array>>`.

File: python/pyarrow/array.pxi [hunk 8]
```diff
@@ -1379,6 +1379,55 @@ cdef class LargeListArray(Array):
         return pyarrow_wrap_array(arr.values())
 
 
+cdef class FixedSizeListArray(Array):
+    """
+    Concrete class for Arrow arrays of a fixed size list data type.
+    """
+
+    @staticmethod
+    def from_arrays(values, int32_t list_size, MemoryPool pool=None):
```
Comment (pitrou) [inline python/pyarrow/array.pxi]: The `pool` argument isn't used here.

File: python/pyarrow/tests/test_convert_builtin.py [hunk 9]
```diff
@@ -148,8 +148,14 @@ def _as_dict_values(xs):
     return six.viewvalues(dct)
 
 
+def _as_array(xs):
```
Comment (pitrou) [inline python/pyarrow/tests/test_convert_builtin.py]: Calling it `_as_np_array` would perhaps be less ambiguous.

File: python/pyarrow/tests/test_schema.py [hunk 10]
```diff
@@ -63,6 +63,7 @@ def test_type_to_pandas_dtype():
         (pa.binary(12), np.object_),
         (pa.string(), np.object_),
         (pa.list_(pa.int8()), np.object_),
+        # (pa.list_(pa.int8(), 2), np.object_),  # TODO needs pandas conversion
```
Comment (pitrou) [inline python/pyarrow/tests/test_schema.py]: Do you plan to do this here or in a separate PR?

File: python/pyarrow/types.pxi [hunk 11]
```diff
@@ -1810,9 +1839,12 @@ cpdef ListType list_(value_type):
     else:
         raise TypeError('List requires DataType or Field')
 
-    list_type.reset(new CListType(_field.sp_field))
-    out.init(list_type)
-    return out
+    if list_size == -1:
+        list_type.reset(new CListType(_field.sp_field))
+    else:
+        list_type.reset(new CFixedSizeListType(_field.sp_field, list_size))
```
Comment (pitrou) [inline python/pyarrow/types.pxi]: Also check that `list_size` is positive?

**Review by pitrou — CHANGES_REQUESTED** [review]
Looks mostly good. Just a few comments here and there.

File: cpp/src/arrow/array.cc (hunk 6 above)
Comment (jorisvandenbossche) [reply cpp/src/arrow/array.cc]: Hmm, that's right. 
But, I don't really see how to construct an array with list_size of 0 in this function, as you would need to know the length of the final array. But this cannot be derived from the length of the values array, since this will have no values for a list_size of 0. So maybe raising an error is fine?

Comment (pitrou) [reply cpp/src/arrow/array.cc]: Yes, raising an error seems reasonable.

**Comment by jorisvandenbossche** [comment]
Thanks for the review @pitrou, updated everything I think.

> Do you plan to do this here or in a separate PR?

I didn't yet look into how complex the conversion to numpy/pandas would be. Will take a quick look, and depending on that might prefer a separate follow-up PR.

**Comment by jorisvandenbossche** [comment]
I updated this PR (rebased + fixed conflicts now the MapArray PR was merged).

I didn't yet add arrow->pandas conversion (but I think that can also be done as follow-up)

**Review by pitrou — APPROVED** [review]
+1. I agree that Pandas conversion can be done later.
